---
title: "03_models"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries
```{r}
library(mctoolsr)
library(plyr) # always load before tidyverse to avoid conflicts with dplyr
library(dplyr)
library(tidyverse) # lazily load all of tidyverse, just in case I want to use it.
library(vegan)
library(MASS)
library(wesanderson)
library(viridis)
library(edgeR)
library(car)
library(performance)
library(lme4)
library(nlme)
library(lmerTest)
library(compositions)

#library(forcats) # for neat ordering of factors when graphing
#library(ggrepel) # for making nice repelling labels

set.seed(10)
```

### Set up input and output directories
```{r Set up input and output directories, include = FALSE}
project.fp <- "/Users/coin8046/Desktop/FiererLabNotebook_copy/SunflowerGxE/r_analyses"
raw_data.fp <- file.path(project.fp, "00_raw.data")
clean_data.fp <- file.path(project.fp, "01_clean.data")
    figures_01_clean.fp <- file.path(clean_data.fp, "figures")
    outputs_01_clean.fp <- file.path(clean_data.fp, "outputs")

model_03.fp <- file.path(project.fp, "03_modeling")
    figures_03.fp <- file.path(model_03.fp, "figures")
    outputs_03.fp <- file.path(model_03.fp, "outputs")

# if (!dir.exists(model_03.fp)) {dir.create(model_03.fp, recursive = TRUE)}
# if (!dir.exists(figures_03.fp)) {dir.create(figures_03.fp, recursive = TRUE)}
# if (!dir.exists(outputs_03.fp)) {dir.create(outputs_03.fp, recursive = TRUE)}
```

### Read in 16S data
```{r}
## cleaned data
input_filt <- readRDS(paste0(outputs_01_clean.fp, "/input_filt.RDS"))

## rarefied data
input_rar <- readRDS(paste0(outputs_01_clean.fp, "/input_filt_rar9k.RDS"))

## relative abundance data
input_relab <- readRDS(paste0(outputs_01_clean.fp, "/input_filt_relab.RDS"))

```

### filter to decrease super rare ASVS
```{r}
low.tmp <- input_filt$data_loaded %>% 
  mutate(ubiq = rowSums(. !=0)) %>% 
  filter(ubiq < 30)

lowASVs <- row.names(low.tmp)

input_filt.rl <- filter_taxa_from_input(input = input_relab, taxa_IDs_to_remove =  lowASVs)
nrow(input_filt.rl$data_loaded)
# < 30:
# 16968 taxa removed
# 10960 remaining

input_filt2 <- filter_taxa_from_input(input = input_filt, taxa_IDs_to_remove =  lowASVs)
# < 30:
# 16968 taxa removed
# 10960 remaining
```

### break datasets into location groups
```{r}
# Mentor, Ralls, Kirkmeyer, Mandan, Pierre
# based on location in ordination (spread out)
input_Mentor <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Mentor")
input_Ralls <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Ralls")
input_Kirkmeyer <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Kirkmeyer")
input_Mandan <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Mandan")
input_Pierre <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Pierre")
input_Carrington <- filter_data(input = input_filt2, filter_cat = "Location", keep_vals = "Carrington")
# input_Mentor
# input_Ralls
# input_Kirkmeyer
# input_Mandan
# input_Pierre
```


### genus level data
```{r}
genus_sum <- summarize_taxonomy(input = input_filt, level = 6, report_higher_tax = FALSE)

genus_sum %>% rowSums() %>% sort(decreasing = TRUE)
# Candidatus Udaeobacter 
# RB41 
# Chthoniobacter
# Candidatus Nitrocosmicus
# Pir4 lineage
# Pirellula 
# Burkholderia-Caballeronia-Paraburkholderia 
# Sphingomonas 
# Candidatus Nitrososphaera 
# Bryobacter
# Streptomyces
# 646 total genera
```


### further filter data 
```{r}
# number samples: 587
# 75% threshold samples: 440... 27879/27928 (49)
# 50% threshold samples: 294... 27603 (325)
# 25% threshold samples: 147... 26183 (1745)
low2.tmp <- input_filt$data_loaded %>% 
  mutate(ubiq = rowSums(. !=0)) %>% 
  filter(ubiq < 147)

low2ASVs <- row.names(low.tmp)

input_filt.u <- filter_taxa_from_input(input = input_filt, taxa_IDs_to_remove =  low2ASVs)
#26183 taxa removed, 1745 ASVs remaining
  
```

#### data distribution vis
```{r}
prefilt <- input_filt$data_loaded

ubiq_pf <- data.frame(ubiq = sort(rowSums(input_filt$data_loaded !=0))) %>%
  rownames_to_column(var = "SampleID") %>%
  ggplot() +
  geom_histogram(aes(x = ubiq), binwidth = 1)

ubiq_f <- data.frame(ubiq = sort(rowSums(input_filt.u$data_loaded !=0))) %>%
  rownames_to_column(var = "SampleID") %>%
  ggplot() +
  geom_histogram(aes(x = ubiq), binwidth = 1)

ggsave(plot = ubiq_pf, filename = paste0(figures_03.fp, "/ubiq_prefilter.png"))
ggsave(plot = ubiq_f, filename = paste0(figures_03.fp, "/ubiq_postfilter.png"))
```

### visualize distributions
#### heat maps to vis ASV distrubution by location
```{r}
hm_pal <- RColorBrewer::brewer.pal(3, "RdBu")
hm_pal <- c("#67A9CF", "#F7F7F7", "#EF8A62")

low.tmp2 <- input_f_samps$data_loaded %>% 
  mutate(ubiq = rowSums(. !=0)) %>% 
  filter(ubiq < 250)

lowASVs2 <- row.names(low.tmp2)

input_filt.hm <- filter_taxa_from_input(input = input_filt, taxa_IDs_to_remove =  lowASVs2)

hm_genus <- summarize_taxonomy(input = input_filt.hm, level = 6, report_higher_tax = FALSE)
hm_asvs <- summarize_taxonomy(input = input_filt.hm, level = 7, report_higher_tax = TRUE)

hm <- plot_ts_heatmap(tax_table = hm_asvs, metadata_map = input_filt.hm$map_loaded, min_rel_abund = .002, type_header = "Location", colors = hm_pal, remove_other = TRUE, scale_by = "taxa") #  
hm$layers[[2]] <- NULL
hm


hm.g <- plot_ts_heatmap(tax_table = hm_genus, metadata_map = input_filt.hm$map_loaded, min_rel_abund = .002, type_header = "Location", colors = hm_pal, remove_other = TRUE, scale_by = "taxa") #  
hm.g$layers[[2]] <- NULL
hm.g

ggsave(hm, filename = paste0(figures_03.fp,"/heatmap.f2.png"), width = 12, height = 15)
ggsave(hm.g, filename = paste0(figures_03.fp,"/heatmap.gf.png"), width = 15, height = 10)
```

#### box plots by location
```{r}
loc.pal <-  c("#D53E4F",  "#FDAE61", "#ABDDA4","#9E0142", "#66C2A5", "#3288BD", "#5E4FA2", "#F46D43","#708090", "#5BA9BC")

low.tmp3 <- input_filt$data_loaded %>% 
  mutate(ubiq = rowSums(. !=0)) %>% 
  filter(ubiq < 500)

lowASVs3 <- row.names(low.tmp3)
topASVs4 <- c("ASV_829", "ASV_781", "ASV_3179", "ASV_4394")

input_filt.bx <- filter_taxa_from_input(input = input_filt, taxa_IDs_to_keep = topASVs4) #taxa_IDs_to_remove =  lowASVs3
nrow(input_filt.bx$data_loaded)

bxplt.df <- input_filt.bx$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "sample_id") %>% 
  inner_join(input_filt.bx$map_loaded, by = c("sample_id" = "BoxCell")) %>% 
  pivot_longer(cols = starts_with("ASV"), names_to = "ASV", values_to = "n.reads")

bxplt3 <- ggplot(bxplt.df, # %>% subset(ASV %in% c("ASV_10")).."ASV_31", "ASV_46" "ASV_10"
                 aes(x = Location, y = n.reads, color = Entry))+
  geom_boxplot()+
  #geom_point(aes(x = Location, y = n.reads, color = Entry))+
  facet_wrap(~ASV, scales = "free_y")+ #nrow = 5, ncol = 4
  theme_bw()+  
  scale_color_manual(values = loc.pal)+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) #size = 12
  
bxplt3

ggsave(filename = paste0(figures_03.fp, "/boxplot_top4.png"), plot = bxplt3, width = 14, height = 7)

#####################
#####################

genus_sum

bxplt.g.df1 <- genus_sum %>% t() %>% as.data.frame() %>% 
  dplyr::select(c("Candidatus Udaeobacter", "Streptomyces", "Burkholderia-Caballeronia-Paraburkholderia"))

names(bxplt.g.df1) <- c("g_Candidatus Udaeobacter", "g_Streptomyces", "g_Burkholderia-Caballeronia-Paraburkholderia")

bxpxlt.g.df <-bxplt.g.df1 %>% 
  rownames_to_column(var = "sample_id") %>% 
  inner_join(input_f_samps$map_loaded, by = c("sample_id" = "BoxCell")) %>% 
  pivot_longer(cols = starts_with("g"), names_to = "genus", values_to = "n.reads")


bxplt4 <- ggplot(bxpxlt.g.df %>% subset(genus %in% c("g_Burkholderia-Caballeronia-Paraburkholderia")), 
                 aes(x = Location, y = n.reads, color = Entry))+
  geom_boxplot()+
  facet_wrap(~genus, scales = "free_y")+ #nrow = 5, ncol = 4
  theme_bw()+  
  scale_color_manual(values = loc.pal)+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) #size = 12
  
bxplt4

####################
####################

bxplt4 <- ggplot(bxpxlt.g.df %>% subset(genus %in% c("g_Candidatus Udaeobacter")), 
                 aes(x = Entry, y = n.reads, color = Location))+
  geom_boxplot()+
  facet_wrap(~genus, scales = "free_y")+ #nrow = 5, ncol = 4
  theme_bw()+  
  #scale_color_manual(values = loc.pal)+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) #size = 12
  
bxplt4

```

### calculate distributions
```{r}
location_tab <- input_filt$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "sample_id") %>% 
  inner_join(input_f_samps$map_loaded, by=c("sample_id" = "BoxCell")) %>% 
  select(-c("Sample_type", "Box", "Cell", "CellB","chlorophyll", "subsample_weight_g",      
  "num","ReadCount.raw","ReadCount.filt")) %>% 
  mutate(numASVs = rowSums(.[,2:27929] !=0)) %>% 
  

location_tab2 <- location_tab %>% 
  select(-c("sample", "plot", "plot_id", "Plot", "Site", "State", "Rep", "Entry", "X", "count", "numASVs")) %>% 
  group_by(Location) %>% 
  column_to_rownames(var = "sample_id") %>% 
  t() %>% as.data.frame()
```


### normalize data
```{r}
#edgeR normalization
input_norm.data.t <- edgeR::normLibSizes(object = input_filt2$data_loaded)
input_norm.data <- edgeR::cpm(y = input_filt2$data_loaded, log = FALSE) %>% as.data.frame()

input_edger <- input_filt2
input_edger$data_loaded <- input_norm.data

# relative abundance data
input_furl <- convert_to_relative_abundances(input = input_filt2)

# rarefied data
input_frar <- single_rarefy(input_filt2, depth = 9000)

# CLR transform - zero inflated normal distrubution
# ancom


# filter

```

### format other normalized data for model input
```{r}

model_tab <- input_frar$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_frar$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

# genus_tab <- genus_sum %>% t() %>% as.data.frame() %>% 
#   rownames_to_column(var = "SampleID") %>% 
#   inner_join(input_f_samps$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
#   # mutate(Entry = as.factor(Entry)) %>% 
#   # mutate(Location = as.factor(Location)) %>% 
#   column_to_rownames(var = "SampleID")
```

### clr transformation for model input
```{r}
# calculate the centered log-ratio using the compositions package

input_clr <- input_filt.rl
clr_relab <-t(clr(ifelse(t(input_clr$data_loaded) > 0, t(input_clr$data_loaded), 1e-7)))
input_clr$data_loaded <- clr_relab

model.clr_tab <- input_clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_filt.rl$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

```

### clr transformation by site
```{r}
# input_Mentor
input_Mentor.clr <- input_Mentor
clr.mentor <-t(clr(ifelse(t(input_Mentor.clr$data_loaded) > 0, t(input_Mentor.clr$data_loaded), 1e-7)))
input_Mentor.clr$data_loaded <- clr.mentor

mentor.clr_tab <- input_Mentor.clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Mentor$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

# input_Ralls
input_Ralls.clr <- input_Ralls
clr.Ralls <-t(clr(ifelse(t(input_Ralls.clr$data_loaded) > 0, t(input_Ralls.clr$data_loaded), 1e-7)))
input_Ralls.clr$data_loaded <- clr.Ralls

ralls.clr_tab <- input_Ralls.clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Ralls$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

# input_Kirkmeyer
input_Kirkmeyer.clr <- input_Kirkmeyer
clr.Kirkmeyer <-t(clr(ifelse(t(input_Kirkmeyer.clr$data_loaded) > 0, t(input_Kirkmeyer.clr$data_loaded), 1e-7)))
input_Kirkmeyer.clr$data_loaded <- clr.Kirkmeyer

Kirkmeyer.clr_tab <- input_Kirkmeyer.clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Kirkmeyer$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

# input_Mandan
input_Mandan.clr <- input_Mandan
Mandan.clr <-t(clr(ifelse(t(input_Mandan.clr$data_loaded) > 0, t(input_Mandan.clr$data_loaded), 1e-7)))
input_Mandan.clr$data_loaded <- Mandan.clr

Mandan.clr_tab <- input_Mandan.clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Mandan$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")


# input_Pierre
input_Pierre.clr <- input_Pierre
Pierre.clr <-t(clr(ifelse(t(input_Pierre.clr$data_loaded) > 0, t(input_Pierre.clr$data_loaded), 1e-7)))
input_Pierre.clr$data_loaded <- Pierre.clr 

Perre.clr_tab <- input_Pierre.clr$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Pierre$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")

# input_Carrington
input_Carrington.clr <- input_Carrington
Carrington.clr <-t(clr(ifelse(t(input_Carrington.clr$data_loaded) > 0, t(input_Carrington.clr$data_loaded), 1e-7)))
input_Carrington.clr$data_loaded <- Carrington.clr


Carrington.clr_tab <- input_Carrington.clr$data_loaded %>% t() %>% as.data.frame() %>% #input_Carrington
  rownames_to_column(var = "SampleID") %>% 
  inner_join(input_Carrington.clr$map_loaded, by=c("SampleID" = "BoxCell")) %>% 
  mutate(Entry = as.factor(Entry)) %>% 
  mutate(Location = as.factor(Location)) %>% 
  column_to_rownames(var = "SampleID")


#### make this into a for loop
```


### test models

#### anova
```{r}
library(ggpubr)
library(rstatix)

# Candidatus Udaeobacter 
# RB41 
# Chthoniobacter
# Candidatus Nitrocosmicus
# Pir4 lineage
# Pirellula 
# Burkholderia-Caballeronia-Paraburkholderia 
# Sphingomonas 
# Candidatus Nitrososphaera 
# Bryobacter
# Streptomyces

# Build the linear model
model  <- lm(Streptomyces ~ Entry*Location,
             data = genus_tab)

# Create a QQ plot of residuals
ggqqplot(residuals(model))
shapiro_test(residuals(model))

# genus_tab %>%
#   group_by(Entry, Location) %>%
#   shapiro_test(Streptomyces)

#check for homogeneity of dispersion
genus_tab %>% levene_test(`Candidatus Udaeobacter`~ Entry*Location)

# anova
m2 <- aov(`Burkholderia-Caballeronia-Paraburkholderia` ~ Entry + Location, data = genus_tab)
summary(m2)

ggqqplot(residuals(m2))
plot(m2, 1)



# tukey test post anova
TukeyHSD(m2, which = "Entry")
```

### linear mixed effects

```{r}
## all locations combined
### everything as a random effect
lme1 <- lmer(ASV_1 ~ 1 + (1|Entry) + (1 |Location), model.clr_tab)
summary(lme1)

lme2 <- lmer(ASV_1 ~ 1 + (1|Entry) + (1 |Rep/Location), model.clr_tab)
summary(lme2)

lme3 <- lmer(ASV_1 ~ 1 + (1|Entry) + (1 |Location/Entry), model.clr_tab)
#== 1|Location + 1|location/genotype
summary(lme3)

### fixed effect

lme4 <- lmer(ASV_1 ~ Entry + (1| Location/Entry), model.clr_tab)
# == Entry + 1| Location + 1|location/genotype
summary(lme4)


~ Entry + (1| Entry/Location)
~ Entry + (1|Location) + Entry|Location + (1| Location/Entry)
# ~Entry*Location + (1|Location) == Entry + Location + Entry:Location


### disease resistance instead of (or in addition to) genotype
~ disease resistance + (1|Entry) + (1|Location)

## add longitude to make location continuous, other climate data?
~ Entry + Longitude + (1| Entry/Location)

## climate data as fixed or random effect
~ Average.of.percent.incidence + (1|cmi.scl) + (1|gdd.scl)  # + (1|Location)

~ cmi.scl + gdd.scl + (1|Entry)


## liklihood ratio test to compare
  ## add in genotype, random effect of location, tells you is this model better


## by location separately
lme1.1 <- lmer(ASV_1 ~ 1 + (1|Entry), data = mentor.clr_tab)
summary(lme1.1)

lme2.1 <- lmer(ASV_1 ~ 1 + (1|Entry) + (1 |Rep), mentor.clr_tab)
summary(lme2.1)
```



### glm / nb glm
```{r}
# negative binomial regression 
# (https://stats.oarc.ucla.edu/r/dae/negative-binomial-regression/)

#input_Mentor.clr
#mentor.clr_tab

glm1 <- glm(ASV_13984 ~ Entry, data = mentor.clr_tab)
summary(glm1)

### check glmm??

#########################
#########################
## old - edger transformed (or otherwise normalized... all data filtered to some threshold)
# m1 <- glm.nb(ASV_7 ~ Entry + Location, data = model_tab)
# summary(m1)
# check_overdispersion(m1)
# m1.g <- glm(Streptomyces  ~ Entry*Location, data = genus_tab)
# summary(m1.g)
# check_overdispersion(m1.g)
# lme4::lmer(Streptomyces  ~ Entry + (1|Location) + (Entry|Location), genus_tab)
```

#### correlations with scelrotinia resistance
```{r}
#ASV_3179, ASV_4394, ASV_781, ASV_829  
cor.test(Carrington.clr_tab$ASV_3179, Carrington.clr_tab$Average.of.percent.incidence)
```

### model loop cor
```{r}
# ### this is to get the final table structure in order
input <- input_Carrington.clr
input_tab <- Carrington.clr_tab
  
## length of the asv list (so the loop knows how many times to loop)
iterations.c = nrow(input$data_loaded)

## 9 genotypes (model drops last) for storing pvals
## 24 entry + location
## 149 with interaction
variables = 2

## make an empty matrix to store results
cor_output <- matrix(ncol=variables, nrow=iterations.c)

test2 <- cor.test(input_tab[,1], input_tab$Average.of.percent.incidence)
test2$estimate

## loop model over all ASVs
for (i in 1:iterations.c){
tmp_cor <- cor.test(x = input_tab[,i], y= input_tab$Average.of.percent.incidence)
tmp_row <- c(tmp_cor$estimate, tmp_cor$p.value)
cor_output[i,] <- tmp_row
}
cor_output <- as.data.frame(cor_output)

names(cor_output) <- c("r", "pval")
rownames(cor_output) <- rownames(input$data_loaded)

cor_output_fdr <- cor_output %>% 
  rownames_to_column(var="asv") %>% 
  mutate(pval.adj = p.adjust(pval, method = "fdr")) %>% 
  mutate(sig = ifelse(pval.adj < 0.001, yes = "***", no = 
                        ifelse(pval.adj < 0.01, yes = "**", no = 
                                 ifelse(pval.adj < 0.05, yes = "*", no = "ns"))))


saveRDS(cor_output_fdr, file = paste0(outputs_03.fp, "/cor_carrington.rl_fdr.RDS"))
# 
# #glm = relative abundance normalized data
# #glm2 = edgeR normalized data
# #glm3 = non-normalized data
```

### model loop glm
```{r}
# ### this is to get the final table structure in order
# input_Mentor.clr, mentor.clr_tab
# input_Ralls.clr, ralls.clr_tab
# input_Kirkmeyer.clr, Kirkmeyer.clr_tab 
# input_Mandan.clr, Mandan.clr_tab
# input_Pierre.clr, Perre.clr_tab

## length of the asv list (so the loop knows how many times to loop)
iterations2 = nrow(input_Pierre.clr$data_loaded)

## 9 genotypes (model drops last) for storing pvals
## 24 entry + location
## 149 with interaction
variables = 20

## make an empty matrix to store results
glm_output <- matrix(ncol=variables, nrow=iterations2)

test2 <- glm(Perre.clr_tab[,1] ~ Entry, data = Perre.clr_tab)
summary(test2)
#unlist(coef(summary(test2))[,3])
#unlist(coef(summary(test2))[,4])

## loop model over all ASVs
for (i in 1:iterations2){
tmp_glm <- glm(Perre.clr_tab[,i] ~ Entry, data = Perre.clr_tab)
tmp_row <- unlist(c(coef(summary(tmp_glm))[,3], coef(summary(tmp_glm))[,4]))
glm_output[i,] <- tmp_row
}
glm_output <- as.data.frame(glm_output)

#names(glm_output) <-
#rownames(glm_output) <-

fdr <-  function(x){
  return (p.adjust(x, method = "fdr"))
  }


glm_output_fdr.r <- data.frame(lapply(glm_output[,11:20],fdr))
glm_output_fdr <- cbind(glm_output[,1:10], glm_output_fdr.r)

names(glm_output_fdr) <- c("Intercept",	"EntryPI 507903",	"EntryPI 531360",	"EntryPI 531361",	"EntryPI 650542",	"EntryPI 650758",	
                         "EntryPI 650808",	"EntryPI 650814",	"EntryPI 650836",	"EntryPI 650839", 
                         "Intercept.pval",	"EntryPI 507903.pval",	"EntryPI 531360.pval",	"EntryPI 531361.pval",	"EntryPI 650542.pval",
                         "EntryPI 650758.pval",	"EntryPI 650808.pval",	"EntryPI 650814.pval",	"EntryPI 650836.pval",	"EntryPI 650839.pval")
rownames(glm_output_fdr) <- rownames(input_Pierre.clr$data_loaded)

saveRDS(glm_output_fdr, file = paste0(outputs_03.fp, "/Pierre.glm.RDS"))
# 
# #glm = relative abundance normalized data
# #glm2 = edgeR normalized data
# #glm3 = non-normalized data
```

# read in saved model outputs and investigate
```{r}
Pierre.glm <- readRDS(file = paste0(outputs_03.fp, "/Pierre.glm.RDS"))
Pierre.glm.filt <- Pierre.glm[apply(Pierre.glm[,12:20]<.07,1,any),] %>% 
  rownames_to_column(var = "ASV")

Kirk.glm <- readRDS(file = paste0(outputs_03.fp, "/Kirkmeyer.glm.RDS"))
Kirk.glm.filt <- Kirk.glm[apply(Kirk.glm[,12:20]<.07,1,any),] %>% 
  rownames_to_column(var = "ASV")

Mandan.glm <- readRDS(file = paste0(outputs_03.fp, "/Mandan.glm.RDS"))
Mandan.glm.filt <- Mandan.glm[apply(Mandan.glm[,12:20]<.07,1,any),] %>% 
  rownames_to_column(var = "ASV")

Mentor.glm <- readRDS(file = paste0(outputs_03.fp, "/mentor.glm.RDS"))
Mentor.glm.filt <- Mentor.glm[apply(Mentor.glm[,12:20]<.07,1,any),] %>% 
  rownames_to_column(var = "ASV")

Ralls.glm <- readRDS(file = paste0(outputs_03.fp, "/Ralls.glm.RDS"))
Ralls.glm.filt <- Ralls.glm[apply(Ralls.glm[,12:20]<.07,1,any),] %>% 
  rownames_to_column(var = "ASV")


all_asv <- Reduce(union, c(Pierre.glm.filt$ASV, Kirk.glm.filt$ASV, Mandan.glm.filt$ASV, Mentor.glm.filt$ASV, Ralls.glm.filt$ASV))
all_asv2 <- c(all_asv, c("ASV_3179", "ASV_4394", "ASV_781", "ASV_829"))

int_asv <- Reduce(intersect, c(Pierre.glm.filt$ASV, Kirk.glm.filt$ASV, Mandan.glm.filt$ASV, Mentor.glm.filt$ASV, Ralls.glm.filt$ASV))

#ASV_3179, ASV_4394, ASV_781, ASV_829  

#"ASV_31",  "ASV_153", "ASV_194",  "ASV_229", "ASV_248", "ASV_280", "ASV_320", "ASV_329", "ASV_460", "ASV_462", "ASV_477", "ASV_565", "ASV_640", "ASV_701", 
```


### model loop lmer
```{r}
## length of the asv list (so the loop knows how many times to loop)
iterations= nrow(input_clr$data_loaded)

## 9 genotypes (model drops last) for storing pvals
## 24 entry + location
## 149 with interaction
variables = 20

## make an empty matrix to store results
lmer_output <- matrix(ncol=variables, nrow=iterations) #iterations

# mm3<- lmer(ASV_440 ~ Entry + (1|Location), data = model.clr_tab)
# summary(mm3)
# anova(mm3)
test <- lmer(model.clr_tab[,1] ~ Entry + (1|Location), data = model.clr_tab)
summary(test)
unlist(coef(summary(test))[,4])
unlist(coef(summary(test))[,5])

## loop model over all ASVs
#iterations
for (i in 1:iterations){
tmp_lmer <- lmer(model.clr_tab[,i] ~ Entry + (1|Location), data = model.clr_tab)
tmp_row <- c(unlist(coef(summary(tmp_lmer))[,4]), unlist(coef(summary(tmp_lmer))[,5]))
lmer_output[i,] <- tmp_row 
}
lmer_output <- as.data.frame(lmer_output)

#names(glm_output) <-
#rownames(glm_output) <- 

fdr <-  function(x){ 
  return (p.adjust(x, method = "fdr"))
  } 


lmer_output_fdr <- data.frame(lapply(lmer_output,fdr)) 

names(lmer_output_fdr) <- c("Intercept",	"EntryPI 507903",	"EntryPI 531360",	"EntryPI 531361",	"EntryPI 650542",	"EntryPI 650758",	
                         "EntryPI 650808",	"EntryPI 650814",	"EntryPI 650836",	"EntryPI 650839", 
                         "Intercept.pval",	"EntryPI 507903.pval",	"EntryPI 531360.pval",	"EntryPI 531361.pval",	"EntryPI 650542.pval",
                         "EntryPI 650758.pval",	"EntryPI 650808.pval",	"EntryPI 650814.pval",	"EntryPI 650836.pval",	"EntryPI 650839.pval")

rownames(lmer_output_fdr) <- rownames(input_clr$data_loaded)

saveRDS(object = lmer_output_fdr, file = paste0(outputs_03.fp, "/lmer.all.output.RDS"))

```

## more palettes
```{r}
library(nationalparkcolors)

pal10 <- c("#D53E4F",  "#FDAE61", "#ABDDA4","#9E0142", "#66C2A5", "#3288BD", "#5E4FA2", "#F46D43","#5BA9BC", "#708090")
pal6 <- c("#D53E4F",  "#ABDDA4","#9E0142", "#5E4FA2", "#F46D43","#708090")

smkmtn6 <- park_palette("SmokyMountains", 6)
```

### latitude by abundance plot per genotype
three lowest resistance genotypes and three highest: PI 531361	PI 650758	PI 650542	PI 650839	PI 507903	PI 650836
```{r}
input_glm <- filter_taxa_from_input(input_filt.rl,taxa_IDs_to_keep =  all_asv2)
# nrow(input_glm$data_loaded)
# 84


gps.loc <- read.csv(paste0(raw_data.fp, '/microbiomeGxEsites_gpscoords.csv'))

glm.df <- input_glm$data_loaded %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "sample_id") %>% 
  inner_join(input_glm$map_loaded, by = c("sample_id" = "BoxCell")) %>% 
  left_join(gps.loc, by = "Location") %>% 
  mutate(lat.ord = as.factor(LAT.y)) %>% 
  pivot_longer(cols = starts_with("ASV"), names_to = "ASV", values_to = "n.reads")

'%!in%' <- function(x,y)!('%in%'(x,y))


#"ASV_153", "ASV_194",  "ASV_229", "ASV_248", "ASV_280", "ASV_320", "ASV_329", "ASV_460", "ASV_462", "ASV_477", "ASV_565", "ASV_640", "ASV_701"
#, "ASV_4394", "ASV_781", "ASV_829", "ASV_3179"
ggplot(data = glm.df %>% subset(ASV %in% c("ASV_781", "ASV_4394", "ASV_3179", "ASV_829") & Entry %!in% c("PI 507903", "PI 531360", "PI 531361", "PI 650758")),
       mapping = aes(x = lat.ord, y = n.reads, color = Entry, group = Entry))+
  geom_jitter()+
  geom_line()+
  facet_wrap(facets = "ASV", nrow = 2,scales = "free_y")+
  scale_color_manual(values = smkmtn6)+
  #scale_color_viridis(begin = .2, end = .8, option = "magma", discrete = TRUE)+
  scale_x_discrete(name ="Latitude (order)", 
                    #labels=c(seq(1,14,1)))+
                   labels=c("Ralls","Lindsborg-dryland", "Lindsborg-irrigated",	"Land Institute",
                            "Burlington",	"Kirkmeyer",	"Brookings",	"Pierre",	"McLaughlin",	"Mandan",
                            "Casselton",	"Grandin",	"Carrington",	"Mentor",	"Velva"))+
  theme_minimal(base_size = 18)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  

```





```{r}
library(glmmTMB)
#https://cran.r-project.org/web/packages/glmmTMB/vignettes/glmmTMB.pdf
#For example, the formula would be 1|block for a random-intercept model or time|block for a model with random variation in slopes through time across groups specified by block. 

# The basic glmmTMB fit — a zero-inflated Poisson model with a single zeroinflation parameter applying to all observations (ziformula~1). (Excluding
# zero-inflation is glmmTMB’s default: to exclude it explicitly, use ziformula~0.)

fit_zipoisson <- glmmTMB(ASV_7 ~ Entry + (1|Location) + (Entry|Location),
data=model.glmmtb_tab,
ziformula=~1,
family=poisson)

fit_zinbinom <- update(fit_zipoisson,family=nbinom2)
fit_zinbinom1 <- update(fit_zipoisson,family=nbinom1)

AICtab(fit_zipoisson,fit_zinbinom,fit_zinbinom1,fit_zinbinom1_bs)
```





